drivers/wifi/esp8266/esp8266.c

```diff
@@ -0,0 +1,844 @@
+
+	esp8266_flag_clear(dev, EDF_STA_CONNECTING);
+}
+
+static int esp8266_mgmt_connect(struct device *dev,
+				struct wifi_connect_req_params *params)
+{
+	struct esp8266_data *data = dev->driver_data;
+	int len;
+
+	if (!net_if_is_up(data->net_iface)) {
+		return -EIO;
+	}
+
+	if (esp8266_flag_is_set(data, EDF_STA_CONNECTED) ||
+	    esp8266_flag_is_set(data, EDF_STA_CONNECTING)) {
+		return -EALREADY;
+	}
+
+	esp8266_flag_set(data, EDF_STA_CONNECTING);
+
+	len = snprintk(data->conn_cmd, sizeof(data->conn_cmd),
+		       "AT+"_CWJAP"=\"");
+	memcpy(&data->conn_cmd[len], params->ssid, params->ssid_length);
+	len += params->ssid_length;
+
+	if (params->security == WIFI_SECURITY_TYPE_PSK) {
+		len += snprintk(&data->conn_cmd[len],
+				sizeof(data->conn_cmd) - len, "\",\"");
+		memcpy(&data->conn_cmd[len], params->psk, params->psk_length);
+		len += params->psk_length;
+	}
+
+	len += snprintk(&data->conn_cmd[len], sizeof(data->conn_cmd) - len,
+			"\"");
+
+	k_work_submit_to_queue(&data->workq, &data->connect_work);
+
+	return 0;
+}
+
+static int esp8266_mgmt_disconnect(struct device *dev)
+{
+	struct esp8266_data *data = dev->driver_data;
+	int ret;
+
+	ret = modem_cmd_send(&data->mctx.iface, &data->mctx.cmd_handler,
+			     NULL, 0, "AT+CWQAP", &data->sem_response,
+			     ESP_CMD_TIMEOUT);
+
+	return ret;
+}
+
+static int esp8266_mgmt_ap_enable(struct device *dev,
+				  struct wifi_connect_req_params *params)
+{
+	char cmd[sizeof("AT+"_CWSAP"=\"\",\"\",xx,x") + WIFI_SSID_MAX_LEN +
+		 WIFI_PSK_MAX_LEN];
+	struct esp8266_data *data = dev->driver_data;
+	int ecn = 0, len, ret;
+
+	ret = modem_cmd_send(&data->mctx.iface, &data->mctx.cmd_handler,
+			     NULL, 0, "AT+"_CWMODE"=3", &data->sem_response,
+			     ESP_CMD_TIMEOUT);
+	if (ret < 0) {
+		LOG_ERR("Failed to enable AP mode, ret %d", ret);
+		return ret;
+	}
+
+	len = snprintk(cmd, sizeof(cmd), "AT+"_CWSAP"=\"");
+	memcpy(&cmd[len], params->ssid, params->ssid_length);
+	len += params->ssid_length;
+
+	if (params->security == WIFI_SECURITY_TYPE_PSK) {
+		len += snprintk(&cmd[len], sizeof(cmd) - len, "\",\"");
+		memcpy(&cmd[len], params->psk, params->psk_length);
+		len += params->psk_length;
+		ecn = 3;
+	} else {
+		len += snprintk(&cmd[len], sizeof(cmd) - len, "\",\"");
+	}
+
+	snprintk(&cmd[len], sizeof(cmd) - len, "\",%d,%d", params->channel,
+		 ecn);
+
+	ret = modem_cmd_send(&data->mctx.iface, &data->mctx.cmd_handler,
+			     NULL, 0, cmd, &data->sem_response,
+			     ESP_CMD_TIMEOUT);
+
+	return ret;
+}
+
+static int esp8266_mgmt_ap_disable(struct device *dev)
+{
+	struct esp8266_data *data = dev->driver_data;
+	int ret;
+
+	ret = modem_cmd_send(&data->mctx.iface, &data->mctx.cmd_handler,
+			     NULL, 0, "AT+"_CWMODE"=1", &data->sem_response,
+			     ESP_CMD_TIMEOUT);
+
+	return ret;
+}
+
+static void esp8266_init_work(struct k_work *work)
+{
+	struct esp8266_data *dev;
+	int ret;
+	static struct setup_cmd setup_cmds[] = {
+		/* turn off echo */
+		SETUP_CMD_NOHANDLE("ATE0"),
```

Comment:
Traditionally, you would start with a plain AT command to ensure the modem was ready.
`SETUP_CMD_NOHANDLE("AT"),`

Reply:
I don't really think that is necessary